core/sector_storage/impl/new_scheduler_impl.hpp

```diff
@@ -0,0 +1,127 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include "sector_storage/scheduler.hpp"
+
+#include <boost/asio/io_context.hpp>
+#include <future>
+#include <mutex>
+#include <unordered_map>
+#include <utility>
+#include "primitives/resources/resources.hpp"
+#include "sector_storage/worker_estimator.hpp"
+#include "storage/buffer_map.hpp"
+
+namespace fc::sector_storage {
+  using primitives::Resources;
+  using storage::BufferMap;
+
+  struct NewTaskRequest {
+    inline NewTaskRequest(const SectorRef &sector,
+                       const TaskType &task_type,
+                       uint64_t priority,
+                       std::shared_ptr<WorkerSelector> sel,
+                       WorkerAction prepare,
+                       WorkerAction work,
+                       ReturnCb cb)
+        : sector(sector),
+          task_type(task_type),
+          priority(priority),
+          sel(std::move(sel)),
+          prepare(std::move(prepare)),
+          work(std::move(work)),
+          cb(std::move(cb)) {
+      const auto &resource_table{primitives::getResourceTable()};
+      const auto resource_iter = resource_table.find(this->task_type);
+      if (resource_iter != resource_table.end()) {
+        need_resources = resource_iter->second.at(sector.proof_type);
+      }
+    };
+
+    SectorRef sector;
+    TaskType task_type;
+    uint64_t priority;
+    std::shared_ptr<WorkerSelector> sel;
+
+    Resources need_resources;
+
+    WorkerAction prepare;
+    WorkerAction work;
+
+    ReturnCb cb;
+  };
+
+  inline bool operator<(const NewTaskRequest &lhs, const NewTaskRequest &rhs) {
+    // priority is intentionally reversed
+    return std::tie(rhs.priority, lhs.task_type, lhs.sector.id.sector)
+           < std::tie(lhs.priority, rhs.task_type, rhs.sector.id.sector);
+  }
+
+  class NewSchedulerImpl : public Scheduler {
```

Comment:
```suggestion
  class EstimateSchedulerImpl : public Scheduler {
```
Please, add comment for the class - it is an improved scheduler with estimator.